src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -423,17 +441,22 @@ private void ExecuteCompilationActionsCore(ImmutableArray<CompilationAnalyzerAct
             }
 
             var addDiagnostic = GetAddDiagnostic(semanticModel.SyntaxTree, analyzer, isSyntaxDiagnostic: false);
+            Func<Diagnostic, bool> isSupportedDiagnostic = d => IsSupportedDiagnostic(analyzer, d);
```

Comment:
There are some cases where it is repeated, but in the most important (most frequently called) cases it is not. In the end, I used a very simply check to determine where to create this capturing delegate: next to calls to `GetAddDiagnostic`.